rtems/Mutex.java

```diff
@@ -49,25 +58,110 @@ public void lock() throws InterruptedException{
 				//how should we prepend here???Doubt the orderRec as it is already present in thisThread.mutexList
 			}
 		  }finally{
-		  	parentLock.unlock();
+			parentLock.unlock();
 		  }
 
 	}
 	public void unlock() throws InterruptedException{
+		Object topMutex=null;
+		RTEMSThread thisThread = (RTEMSThread)Thread.currentThread();
+		RTEMSThread candidateThr;
+		int stepdownPri;
 		parentLock.lock();
 		//proper step down of priority.
 		//remove eligible candidate thread from this.waitQueue
 		//set the state of that thread to Params.RUNNABLE
 		//signalAll()
 		//unlockparentLock.lock()
+		try{
+			//1.Assertion Check
+			topMutex = thisThread.mutexOrderList.get(0);
+			if(topMutex!=this){
+				//assertion error for strict order mutexes
+			}
+			//Not handling the nesting case right now.
+			//2. Remove top element from mutexOrderList
+			//mutexOrderList is acting as chaincontrol structure for us.
+			//Right now avoiding complications
+			topMutex = thisThread.mutexOrderList.remove(0);
+			//3. Initialize the mutex orderList
+			//this.orderRec.node.init();
+			//4. stepdown of priority for this thread
+			stepdownPri = this.priorityBefore;
+
+			thisThread.setPriority(stepdownPri);
+			//5. Re-enqueue if thread is waiting
+			reEnqueue();
+			holder = null;
+			candidateThr = (RTEMSThread)waitQueue.poll();
+			if(candidateThr != null){
+				candidateThr.state = Thread.State.RUNNABLE;
+				//Logically only candidate will go through lock and rest will again get queued up
+				cv1.notifyAll();
+			}
+
+		}finally{
+			parentLock.unlock();
+		}
+	}
+
+	public boolean priorityRaiseFilter(int priority){
+		int holderPriority = holder.getPriority();
+		if(priority < holderPriority){
+			return true;
+		}
+		return false;
 	}
-	public void updatePriority()
-	{
 
+	public void updatePriority(int priority)
+	{
+		holder.currentPriority = priority;
+		holder.setPriority(priority);
```

Comment:
What's the difference between the RTEMSThread current priority and the thread object's priority? Will they ever be different?